File: pdf_toolkit/image_context/ports.py
"""image context — ports."""

from __future__ import annotations

from abc import ABC, abstractmethod

from pdf_toolkit.shared_kernel import PerceptualScore, ResampleFilter


class ImageCodec(ABC):
    @abstractmethod
    def decode(self, data: bytes): ...

    @abstractmethod
    def resize(self, img, width: int, height: int, resample: ResampleFilter): ...

    @abstractmethod
    def encode(self, img, fmt: str, quality: int) -> bytes: ...


class QualityMeter(ABC):
    @abstractmethod
    def score(self, original: bytes, candidate: bytes) -> PerceptualScore: ...


class UnsupportedFormat(ValueError):
    """Format has no lossless optimization path."""


class LosslessOptimizer(ABC):
    """Shrink an encoded image with bit-exact decoded pixels (DOMAIN.md image §Lossless)."""

    @abstractmethod
    def optimize(self, data: bytes, fmt: str) -> bytes: ...

File: rust/src/image_context/ports.rs
//! image context — ports (traits).

use super::domain::ImageFormat;
use crate::pdf_context::domain::ResampleFilter;
use crate::shared_kernel::PerceptualScore;

/// Decodes, resizes and re-encodes raster images. The decoded image type is
/// adapter-specific (an associated type), mirroring Python's duck-typed port.
pub trait ImageCodec {
    type Image;
    fn decode(&self, data: &[u8]) -> Result<Self::Image, String>;
    fn resize(
        &self,
        img: &Self::Image,
        width: u32,
        height: u32,
        resample: ResampleFilter,
    ) -> Result<Self::Image, String>;
    fn encode(&self, img: &Self::Image, format: ImageFormat, quality: u32) -> Result<Vec<u8>, String>;
}

pub trait QualityMeter {
    fn score(&self, original: &[u8], candidate: &[u8]) -> Result<PerceptualScore, String>;
}
